crud/src/main/java/com/redhat/lightblue/crud/valuegenerators/GeneratedFields.java

```diff
@@ -0,0 +1,167 @@
+/*
+ Copyright 2013 Red Hat, Inc. and/or its affiliates.
+
+ This file is part of lightblue.
+
+ This program is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.redhat.lightblue.crud.valuegenerators;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.redhat.lightblue.metadata.FieldCursor;
+import com.redhat.lightblue.metadata.FieldTreeNode;
+import com.redhat.lightblue.metadata.FieldConstraint;
+import com.redhat.lightblue.metadata.ValueGenerator;
+import com.redhat.lightblue.metadata.SimpleField;
+import com.redhat.lightblue.metadata.EntityMetadata;
+import com.redhat.lightblue.metadata.constraints.IdentityConstraint;
+import com.redhat.lightblue.metadata.constraints.RequiredConstraint;
+import com.redhat.lightblue.crud.Factory;
+import com.redhat.lightblue.util.JsonDoc;
+import com.redhat.lightblue.util.KeyValueCursor;
+import com.redhat.lightblue.util.MutablePath;
+import com.redhat.lightblue.util.Path;
+import com.redhat.lightblue.extensions.valuegenerator.ValueGeneratorSupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+/**
+ * Initializes generated fields if they are not already initialized.
+ *
+ * If a generated field is required but it is not in the document, or
+ * it is in the document with a null value, it is inserted into the
+ * document and initialized.
+ *
+ * If a generated field is not required, but it exists in the document
+ * with null value, it is initialized.
```

Comment:
The line between a field not existing, and existing with a null value is
somewhat blurred. In many places, if a field doesn't exist, getting its
value will return null. That's why I considered null and non-existence as
uninitialized.

On Tue, Sep 1, 2015 at 9:03 AM, Marek <notifications@github.com> wrote:

> In
> crud/src/main/java/com/redhat/lightblue/crud/valuegenerators/GeneratedFields.java
> <https://github.com/lightblue-platform/lightblue-core/pull/443#discussion_r38429327>
> :
>
> > +import com.redhat.lightblue.util.Path;
> > +import com.redhat.lightblue.extensions.valuegenerator.ValueGeneratorSupport;
> > +import org.slf4j.Logger;
> > +import org.slf4j.LoggerFactory;
> > +
> > +import java.util.List;
> > +
> > +/**
> > + * Initializes generated fields if they are not already initialized.
> > + *
> > + * If a generated field is required but it is not in the document, or
> > + * it is in the document with a null value, it is inserted into the
> > + * document and initialized.
> > + *
> > + * If a generated field is not required, but it exists in the document
> > + * with null value, it is initialized.
>
> It is allowed to set required field to null. It's a valid value, I'm not
> sure if it should trigger initialization. For example, we can have a
> required lastUpdateDate set to null, because it has not been updated yet.
> It's not a very good example, since usually lastUpdateDate matches creation
> date before first update, but I don't have a better one.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-core/pull/443/files#r38429327>
> .
>